gobblin-core/src/main/java/org/apache/gobblin/converter/avro/JsonElementConversionWithAvroSchemaFactory.java

```diff
@@ -210,4 +216,70 @@ public Schema schema() {
       return this.schema;
     }
   }
-}
+
+  public static class UnionConverter extends ComplexConverter {
+    private final List<Schema> schemas;
+    private final List<JsonElementConverter> converters;
+    private final Schema schemaNode;
+
+    public UnionConverter(String fieldName, boolean nullable, String sourceType, Schema schemaNode,
+        WorkUnitState state, List<String> ignoreFields) throws UnsupportedDateTypeException {
+      super(fieldName, nullable, sourceType);
+      this.schemas = schemaNode.getTypes();
+      converters = new ArrayList<>();
+      for(Schema schema: schemas) {
+        converters.add(getConvertor(fieldName, schema.getType().getName(), schemaNode, state, isNullable(), ignoreFields));
+      }
+      this.schemaNode = schemaNode;
+    }
+
+    @Override
+    Object convertField(JsonElement value) {
+       for(JsonElementConverter converter: converters)
+       {
+         try {
+           Object o = converter.convert(value);
+           return o;
+         } catch (Exception e){}
+       }
+       throw new RuntimeException(StringFormatter.format("Cannot convert %s to avro using schema %s", value.getAsString(), schemaNode.toString()).toString());
+    }
+
+    @Override
+    public Schema.Type getTargetType() {
+      if(schemas.size() == 2 && isNullable())
+      {
+        if(schemas.get(0).getType() == Schema.Type.NULL) {
+          return schemas.get(1).getType();
+        } else {
+          return schemas.get(0).getType();
+        }
+      }
+      return Schema.Type.UNION;
+    }
+
+    @Override
+    public Schema schema() {
+      if(schemas.size() == 2 && isNullable())
```

Comment:
Sorry, I did not understand why this check is required.
Also, start braces should be in this line, not in the next line

Reply:
This is a requirement for avro, if it's a union of null and XXX type, it's a nullable XXX type in avro.

Reply:
What should be the schema, if it is a union of null and XXX type and YYY type?

Reply:
And then it's still a union

Reply:
Ok